pkg/federate/kubefed/resource_test.go

```diff
@@ -0,0 +1,389 @@
+	var (
+		resource *corev1.Pod
+		expected *unstructured.Unstructured
+		scheme   *runtime.Scheme
+	)
+
+	BeforeEach(func() {
+		resource = newPod("test-pod", "nginx")
+		scheme = runtime.NewScheme()
+	})
+
+	AfterEach(func() {
+		resource = nil
+		expected = nil
+		scheme = nil
+	})
+
+	When("the resource type scheme has NOT been added to the type registry", func() {
+		When("failing to convert to Unstructured", func() {
+			It("should throw an error", func() {
+				var expectedType *UnstructuredConversionError
+				_, err := createFederatedResource(scheme, resource)
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(BeAssignableToTypeOf(expectedType))
+			})
+
+			It("should return a null expected", func() {
+				expected, _ = createFederatedResource(scheme, resource)
+				Expect(expected).To(BeNil())
+			})
+		})
+	})
+
+	When("the resource type scheme has been added to the type registry", func() {
+		BeforeEach(func() {
+			err := corev1.AddToScheme(scheme)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("should not modify the input resource but copy it", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(resource).ToNot(BeIdenticalTo(expected))
+		})
+
+		It("should return no error on success", func() {
+			_, err := createFederatedResource(scheme, resource)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		It("the returned resource's Kind should be of Federated<kind>", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			expectedKind := federatedKindPrefix + resource.GroupVersionKind().Kind
+			Expect(expected.GroupVersionKind().Kind).
+				To(Equal(expectedKind))
+		})
+
+		It("the returned resource's Group should be the default KubeFed FederatedGroup", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Group).
+				To(Equal(kubefedopt.DefaultFederatedGroup))
+		})
+
+		It("the returned resource's Version should be the default KubeFed FederatedVersion", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GroupVersionKind().Version).
+				To(Equal(kubefedopt.DefaultFederatedVersion))
+		})
+
+		It("the returned resource's name should be the same as the input's object", func() {
+			expected, _ = createFederatedResource(scheme, resource)
+			Expect(expected.GetName()).
+				To(Equal(resource.GetName()))
+		})
+
+		It("the returned resource's template should be the input's unstructured", func() {
+			// This test implicitly tests that the nested hierarchy of fields exists
+			expected, _ = createFederatedResource(scheme, resource)
+
+			mapObj, found, err := getTemplateField(expected)
+			targetResource := &unstructured.Unstructured{}
+
+			convErr := scheme.Convert(resource, targetResource, nil)
+			Expect(convErr).ToNot(HaveOccurred())
+			removeUnwantedFields(targetResource)
+
+			Expect(err).ToNot(HaveOccurred())
+			Expect(found).To(BeTrue(), "Hierarchy of fields missing!")
```

Comment:
Can we change this to _"The template field is missing!"_?

Reply:
That would be technically incorrect as it's not only the template field but the whole hierarchy until the template field.

Reply:
Not sure I follow -  you're specifically getting the nested "templates" field here right?